### PR TITLE
Fix sync broadcast and poll for transaction status

### LINKED ISSUE #2324: Yoda should use sync broadcast and poll for transaction status
Some validators report to run commit timeout issues. I think polling here makes more sense.

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed: #2324 

## Implementation details
1. Fix broadcast from commit to sync
2. Polling for transaction status

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [x] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [x] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

## If this PR doesn't need to be merged into v1.0.6 please remove this option:

- [ ] The pull request has been cherry-picked to `rc/1.0.6`

### CODE REVIEW

File: chain/yoda/execute.go [hunk 1]
```diff
@@ -51,16 +55,31 @@ func SubmitReport(c *Context, l *Logger, id otypes.RequestID, reps []otypes.RawR
...
+		if err != nil {
+			l.Error(":exploding_head: Failed to broadcast tx with error: %s", err.Error())
+			time.Sleep(SleepTime)
+			continue
+		}
+		for start := time.Now(); time.Since(start) < BroadCastTimeout; {
+			time.Sleep(SleepTime)
+			txRes, err := utils.QueryTx(cliCtx, res.TxHash)
+			if err != nil {
+				l.Error(":exploding_head: Failed to query tx with error: %s", err.Error())
```
Comment (sorawit) [inline chain/yoda/execute.go]: Isn't this expected? Why do we log error with exploding head?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/yoda/execute.go [hunk 2]
```diff
@@ -51,16 +55,31 @@ func SubmitReport(c *Context, l *Logger, id otypes.RequestID, reps []otypes.RawR
...
+				continue
+			}
+			if txRes.Code != 0 {
+				l.Error(":exploding_head: Tx returned nonzero code %d with log %s, tx hash: %s", txRes.Code, txRes.RawLog, txRes.TxHash)
+				break
```
Comment (taobun) [inline chain/yoda/execute.go]: return

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/yoda/execute.go [hunk 3]
```diff
@@ -50,17 +53,30 @@ func SubmitReport(c *Context, l *Logger, id otypes.RequestID, reps []otypes.RawR
...
+		if err == nil {
+			break
+		}
 		l.Error(":exploding_head: Failed to broadcast tx with error: %s", err.Error())
-		return
+		time.Sleep(SleepTime)
 	}
-	if res.Code != 0 {
-		l.Error(":exploding_head: Tx returned nonzero code %d with log %s, tx hash: %s", res.Code, res.RawLog, res.TxHash)
+	for start := time.Now(); time.Since(start) < c.broadcastTimeout; {
```
Comment (taobun) [inline chain/yoda/execute.go:71]: What happens after trying until MaxTry and cannot send tx (got error every try) We need to query on empty tx response?

File: chain/yoda/execute.go [hunk 4]
```diff
@@ -50,17 +53,30 @@ func SubmitReport(c *Context, l *Logger, id otypes.RequestID, reps []otypes.RawR
...
+			l.Debug("Failed to query tx with error: %s", err.Error())
+			continue
+		}
+		if txRes.Code != 0 {
+			l.Error(":exploding_head: Tx returned nonzero code %d with log %s, tx hash: %s", txRes.Code, txRes.RawLog, txRes.TxHash)
+			return
+		}
+		l.Info(":smiling_face_with_sunglasses: Successfully broadcast tx with hash: %s", txRes.TxHash)
 		return
 	}
```
Comment (taobun) [inline chain/yoda/execute.go:84]: Need to log if cannot find tx if timeout?

File: chain/yoda/run.go [hunk 5]
```diff
@@ -107,11 +111,13 @@ func runCmd(c *Context) *cobra.Command {
 	cmd.Flags().String(flagExecutor, "", "executor name and url for executing the data source script")
 	cmd.Flags().String(flags.FlagGasPrices, "", "gas prices for report transaction")
 	cmd.Flags().String(flagLogLevel, "info", "set the logger level")
+	cmd.Flags().String(flagBroadcastTimeout, "", "The time that Yoda will wait for tx commit")
```
Comment (taobun) [inline chain/yoda/run.go]: Add `10s` as default value

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/scripts/start_yoda.sh [hunk 6]
```diff
@@ -11,6 +11,9 @@ yoda config validator $(bandcli keys show $1 -a --bech val --keyring-backend tes
 # setup execution endpoint
 yoda config executor "docker:bandprotocol/runtime:1.0.1"
 
+# setup broadcast-timeout  to yoda config
```
Comment (taobun) [inline chain/scripts/start_yoda.sh]: Space

File: chain/scripts/start_yoda.sh [hunk 7]
```diff
@@ -11,6 +11,9 @@ yoda config validator $(bandcli keys show $1 -a --bech val --keyring-backend tes
...
+yoda config broadcast-timeout "10s"
```
Comment (taobun) [inline chain/scripts/start_yoda.sh]: @sorawit 10 seconds is the default of BroadcastTxCommit if we want to solve this problem how much time is a recommended option?

File: chain/yoda/execute.go [hunk 8]
```diff
@@ -50,17 +53,37 @@ func SubmitReport(c *Context, l *Logger, id otypes.RequestID, reps []otypes.RawR
...
+	var res sdk.TxResponse
+	found := false
+	for try := 1; try <= MaxTry; try++ {
+		l.Info("Try to broadcast: %d/%d", try, MaxTry)
+		res, err = cliCtx.BroadcastTxSync(out)
+		if err == nil {
+			found = true
+			break
+		}
 		l.Error(":exploding_head: Failed to broadcast tx with error: %s", err.Error())
```
Comment (taobun) [inline chain/yoda/execute.go]: We don't want to see this log to be critical error. There are 2 options that I think it make sense
1. Change icon to ⚠️ 
2. Change log level to debug.

File: chain/yoda/execute.go [hunk 9]
```diff
@@ -50,17 +53,37 @@ func SubmitReport(c *Context, l *Logger, id otypes.RequestID, reps []otypes.RawR
 		l.Error(":exploding_head: Failed to build tx with error: %s", err.Error())
 		return
 	}
-
-	res, err := cliCtx.BroadcastTxCommit(out)
-	if err != nil {
+	var res sdk.TxResponse
+	found := false
+	for try := 1; try <= MaxTry; try++ {
+		l.Info("Try to broadcast: %d/%d", try, MaxTry)
```
Comment (taobun) [inline chain/yoda/execute.go]: Try to broadcast report transaction(%d/%d)

Comment (taobun) [reply chain/yoda/execute.go]: add some emoji too.

File: chain/yoda/execute.go [hunk 10]
```diff
@@ -50,17 +53,37 @@ func SubmitReport(c *Context, l *Logger, id otypes.RequestID, reps []otypes.RawR
...
+		}
+		if txRes.Code != 0 {
+			l.Error(":exploding_head: Tx returned nonzero code %d with log %s, tx hash: %s", txRes.Code, txRes.RawLog, txRes.TxHash)
+			return
+		}
+		l.Info(":smiling_face_with_sunglasses: Successfully broadcast tx with hash: %s", txRes.TxHash)
 		return
 	}
-	l.Info(":smiling_face_with_sunglasses: Successfully broadcast tx with hash: %s", res.TxHash)
+	l.Info("Can't find tx hash: %s before timeout", res.TxHash)
```
Comment (taobun) [inline chain/yoda/execute.go]: ❓ Cannot get transaction response from hash: %s transaction might be included in the next few blocks or check your node's health.

File: chain/yoda/execute.go [hunk 11]
```diff
@@ -50,17 +53,37 @@ func SubmitReport(c *Context, l *Logger, id otypes.RequestID, reps []otypes.RawR
...
+		l.Error(":exploding_head: Can't try to broadcast more than %d try", MaxTry)
 		return
 	}
-	if res.Code != 0 {
-		l.Error(":exploding_head: Tx returned nonzero code %d with log %s, tx hash: %s", res.Code, res.RawLog, res.TxHash)
+	for start := time.Now(); time.Since(start) < c.broadcastTimeout; {
+		time.Sleep(SleepTime)
+		txRes, err := utils.QueryTx(cliCtx, res.TxHash)
+		if err != nil {
+			l.Debug("Failed to query tx with error: %s", err.Error())
```
Comment (taobun) [inline chain/yoda/execute.go]: emoji!

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/yoda/execute.go [hunk 12]
```diff
@@ -2,19 +2,23 @@ package yoda
...
 
 	"github.com/bandprotocol/bandchain/chain/app"
 	otypes "github.com/bandprotocol/bandchain/chain/x/oracle/types"
 	rpcclient "github.com/tendermint/tendermint/rpc/client"
 )
 
 var (
-	cdc = app.MakeCodec()
+	cdc       = app.MakeCodec()
+	MaxTry    = 5
```
Comment (taobun) [inline chain/yoda/execute.go]: Move to config

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/yoda/run.go [hunk 13]
```diff
@@ -107,11 +111,13 @@ func runCmd(c *Context) *cobra.Command {
 	cmd.Flags().String(flagExecutor, "", "executor name and url for executing the data source script")
 	cmd.Flags().String(flags.FlagGasPrices, "", "gas prices for report transaction")
 	cmd.Flags().String(flagLogLevel, "info", "set the logger level")
+	cmd.Flags().String(flagBroadcastTimeout, "10s", "The time that Yoda will wait for tx commit")
```
Comment (sorawit) [inline chain/yoda/run.go]: 30s

File: chain/yoda/execute.go [hunk 14]
```diff
@@ -2,19 +2,23 @@ package yoda
...
+	SleepTime = 3 * time.Second
```
Comment (sorawit) [inline chain/yoda/execute.go]: Change this config and rename it to `RPCPollInterval`

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/yoda/run.go [hunk 15]
```diff
@@ -107,11 +120,17 @@ func runCmd(c *Context) *cobra.Command {
 	cmd.Flags().String(flagExecutor, "", "executor name and url for executing the data source script")
 	cmd.Flags().String(flags.FlagGasPrices, "", "gas prices for report transaction")
 	cmd.Flags().String(flagLogLevel, "info", "set the logger level")
+	cmd.Flags().String(flagBroadcastTimeout, "30s", "The time that Yoda will wait for tx commit")
+	cmd.Flags().String(flagRPCPollInterval, "3s", "The duration of rpc poll interval")
+	cmd.Flags().String(flagMaxTry, "5", "Maximum times of try to report transaction")
```
Comment (taobun) [inline chain/yoda/run.go]: cmd.Flags().Uint64(...)

File: chain/yoda/main.go [hunk 16]
```diff
@@ -15,19 +15,25 @@ import (
...
-	Executor  string `mapstructure:"executor"`   // Executor name and URL (example: "Executor name:URL")
+	ChainID          string `mapstructure:"chain-id"`          // ChainID of the target chain
+	NodeURI          string `mapstructure:"node"`              // Remote RPC URI of BandChain node to connect to
+	Validator        string `mapstructure:"validator"`         // The validator address that I'm responsible for
+	GasPrices        string `mapstructure:"gas-prices"`        // Gas prices of the transaction
+	LogLevel         string `mapstructure:"log-level"`         // Log level of the logger
+	Executor         string `mapstructure:"executor"`          // Executor name and URL (example: "Executor name:URL")
+	BroadcastTimeout string `mapstructure:"broadcast-timeout"` // The time that Yoda will wait for tx commit
+	RPCPollInterval  string `mapstructure:"rpc-poll-interval"` // The duration of rpc poll interval
+	MaxTry           string `mapstructure:"max-try"`           // Maximum times of try to report transaction
```
Comment (taobun) [inline chain/yoda/main.go]: uint64

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/scripts/start_yoda.sh [hunk 17]
```diff
@@ -11,6 +11,15 @@ yoda config validator $(bandcli keys show $1 -a --bech val --keyring-backend tes
 # setup execution endpoint
 yoda config executor "docker:bandprotocol/runtime:1.0.1?timeout=10s"
```
Comment (taobun) [inline chain/scripts/start_yoda.sh]: Can you change this to 1.0.2?

File: chain/yoda/run.go [hunk 18]
```diff
@@ -107,11 +116,17 @@ func runCmd(c *Context) *cobra.Command {
 	cmd.Flags().String(flagExecutor, "", "executor name and url for executing the data source script")
 	cmd.Flags().String(flags.FlagGasPrices, "", "gas prices for report transaction")
 	cmd.Flags().String(flagLogLevel, "info", "set the logger level")
+	cmd.Flags().String(flagBroadcastTimeout, "30s", "The time that Yoda will wait for tx commit")
+	cmd.Flags().String(flagRPCPollInterval, "3s", "The duration of rpc poll interval")
+	cmd.Flags().Uint64(flagMaxTry, 5, "Maximum times of try to report transaction")
```
Comment (taobun) [inline chain/yoda/run.go]: The maximum number of tries to submit a report transaction.

File: chain/yoda/main.go [hunk 19]
```diff
@@ -15,19 +15,25 @@ import (
...
-	Executor  string `mapstructure:"executor"`   // Executor name and URL (example: "Executor name:URL")
+	ChainID          string `mapstructure:"chain-id"`          // ChainID of the target chain
+	NodeURI          string `mapstructure:"node"`              // Remote RPC URI of BandChain node to connect to
+	Validator        string `mapstructure:"validator"`         // The validator address that I'm responsible for
+	GasPrices        string `mapstructure:"gas-prices"`        // Gas prices of the transaction
+	LogLevel         string `mapstructure:"log-level"`         // Log level of the logger
+	Executor         string `mapstructure:"executor"`          // Executor name and URL (example: "Executor name:URL")
+	BroadcastTimeout string `mapstructure:"broadcast-timeout"` // The time that Yoda will wait for tx commit
+	RPCPollInterval  string `mapstructure:"rpc-poll-interval"` // The duration of rpc poll interval
+	MaxTry           uint64 `mapstructure:"max-try"`           // Maximum times of try to report transaction
```
Comment (taobun) [inline chain/yoda/main.go]: The maximum number of tries to submit a report transaction

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**